pkg/sidecar/fanout/fanout_handler.go

```diff
@@ -63,103 +64,56 @@ type forwardMessage struct {
 
-	go handler.foreverDispatch(stopCh)
 	return handler
 }
 
 func createReceiverFunction(f *Handler) func(buses.ChannelReference, *buses.Message) error {
 	return func(_ buses.ChannelReference, m *buses.Message) error {
-		done := make(chan error)
-		msg := &forwardMessage{
-			msg:  m,
-			done: done,
-		}
-
-		select {
-		case f.receivedMessages <- msg:
-			// Continue to next select.
-		default:
-			f.logger.Debug("Unable to add message to channel, dropping it.", zap.Any("msg", msg))
-			return errors.New("unable to add message to channel, dropping it")
-		}
-
-		select {
-		case possibleErr := <-done:
-			f.logger.Debug("Responding", zap.Any("possibleErr", possibleErr))
-			return possibleErr
-		case <-time.After(f.timeout):
-			f.logger.Debug("Timeout waiting for dispatch")
-			return errors.New("timeout waiting for dispatch")
-		}
+		return f.dispatch(m)
 	}
 }
 
 func (f *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	f.receiver.HandleRequest(w, r)
 }
 
-// Stop stops the background goroutine. Once this is called, this Handler will no longer properly
-// Serve HTTP traffic.
-func (f *Handler) Stop() {
-	f.stopCh <- struct{}{}
-}
-
-// foreverDispatch dispatches received messages in an infinite loop. It exits only when
-// Handler.stopCh receives a message. It is meant to be run as a goroutine.
-func (f *Handler) foreverDispatch(stopCh <-chan struct{}) {
-	for {
-		select {
-		case msg := <-f.receivedMessages:
-			f.dispatch(msg)
-		case <-stopCh:
-			f.logger.Info("Fanout dispatch thread stopping.")
-			return
-		}
+// dispatch takes the request, fans it out to each subscription in f.config. If all the fanned out
+// requests return successfully, then return nil. Else, return an error.
+func (f *Handler) dispatch(msg *buses.Message) error {
+	errorCh := make(chan error, len(f.config.Subscriptions))
+	for _, sub := range f.config.Subscriptions {
+		go func(s duckv1alpha1.ChannelSubscriberSpec) {
+			errorCh <- f.makeFanoutRequest(*msg, s)
+		}(sub)
 	}
-}
 
-// ServeHTTP takes the request, fans it out to each subscription in f.config. If all the fanned out
-// requests return successfully, then return successfully. Else, return failure.
-func (f *Handler) dispatch(msg *forwardMessage) {
-	msg.done <- func() error {
-		errorCh := make(chan error, len(f.config.Subscriptions))
-		for _, sub := range f.config.Subscriptions {
-			go func(s duckv1alpha1.ChannelSubscriberSpec) {
-				errorCh <- f.makeFanoutRequest(*msg.msg, s)
-			}(sub)
-		}
-
-		for range f.config.Subscriptions {
-			select {
-			case err := <-errorCh:
-				if err != nil {
-					f.logger.Error("Fanout had an error", zap.Error(err))
-					return err
-				}
-			case <-time.After(f.timeout):
-				f.logger.Error("Fanout timed out")
-				return errors.New("fanout timed out")
+	for range f.config.Subscriptions {
+		select {
+		case err := <-errorCh:
+			if err != nil {
+				f.logger.Error("Fanout had an error", zap.Error(err))
+				return err
```

Comment:
Just to confirm the intended behavior: the first failure causes an immediate error response without waiting for the rest of the requests to complete. Those remaining requests execute fully and send their errors to the channel, but the channel is ignored and will be garbage collected after the last request's goroutine completes.

Reply:
Yes, that is the intended behavior. 

I could be convinced we need to wait for all requests to finish before responding. I haven't found a good requirement for it (yet), so went with the method that gets responses more quickly.